packages/transaction-preview/src/index.ts

```diff
@@ -0,0 +1,259 @@
+
+export * from './types.js'
+
+const approved$ = new Subject<boolean>()
+let options: TransactionPreviewInitOptions
+let app: TransactionPreview
+
+export const setContainerElement = (containerElement: string): void => {
+  if (containerElement) {
+    const error = validateSetContainerEl(containerElement)
+
+    if (error) {
+      throw error
+    }
+  }
+  options.containerElement = containerElement
+}
+
+const destroyApp = () => {
+  app.$destroy()
+}
+
+const simTransactions = (
+  txs: [TransactionObject]
+): Promise<SimPlatformResponse> => {
+  if (txs) {
+    const error = validateSimTransactions(txs)
+
+    if (error) {
+      throw error
+    }
+  }
+  return simulateTransactions(options, txs)
+}
+
+const handleRequireApproval = async (
+  app: TransactionPreview,
+  fullProviderRequest: EIP1193Provider['request'],
+  req: {
+    method: string
+    params?: Array<unknown>
+  }
+) => {
+  const approved = await firstValueFrom(approved$)
+  app.$destroy()
+  if (!approved) {
+    throw new ProviderRpcError({
+      code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
+      message: 'User rejected the transaction'
+    })
+  }
+  fullProviderRequest(req)
+}
+
+export const patchProvider = (
+  walletProvider: PatchedEIP1193Provider
+): PatchedEIP1193Provider => {
+  if (!walletProvider) {
+    throw new Error(
+      `An EIP 1193 wallet provider is required to preform patching and 
+      watch for transactions e.g. an injected wallet using window.ethereum`
+    )
+  }
+  if (walletProvider.simPatched) return walletProvider as PatchedEIP1193Provider
+
+  const fullProviderRequest = walletProvider.request
+  const patchedProvider = walletProvider as PatchedEIP1193Provider
+  const request: EIP1193Provider['request'] = async (req: {
+    method: string
+    params?: Array<unknown>
+  }): Promise<any> => {
+    if (
+      req.method === 'eth_sendTransaction' &&
+      req.params &&
+      req.params.length
+    ) {
+      const transactionParams =
+        req.params as EthSignTransactionRequest['params']
+      try {
+        const preview = await simulateTransactions(options, transactionParams)
+        if (preview.status !== 'simulated') {
+          // If transaction simulation was unsuccessful do not create DOM el
+          return
+        }
+        if (app) app.$destroy()
+        app = mountTransactionPreview(preview)
+        options.requireTransactionApproval
+          ? handleRequireApproval(app, fullProviderRequest, req)
+          : fullProviderRequest(req).then((hash: unknown) => {
+              hash && app.$destroy()
+            }).catch(() => app.$destroy())
+      } catch (e) {
+        throw new ProviderRpcError({
+          code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
+          message: `User rejected the transaction: ${e}`
+        })
+      }
+    } else {
+      return fullProviderRequest(req)
+    }
+  }
+  patchedProvider.request = request
+  patchedProvider.simPatched = true
```

Comment:
I wonder if we need to wrap the whole call in a try catch statement? I vaguely remember that some providers are set to read only, so we may have trouble with some wallets when setting the `simPatched` value, but I could be wrong on that. We could log a warning to the console in the case that we could not patch the provider.

Reply:
Ah good call, I think Binance wallet is one of those!

Reply:
@aaronbarnardsound I added a try around the reassigning of the request property.
```  
try {
    patchedProvider.request = request
    patchedProvider.simPatched = true
  } catch (err) {
    console.error(
      `There was an error patching the passed in wallet provider. 
      The provider may be read only and may be incompatible with Transaction Preview`
    )
  }
  ```
  Is that along the lines of what you had in mind?

Reply:
Yep!